request.go

```diff
@@ -0,0 +1,184 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tspclient
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// MessageImprint contains the hash of the datum to be time-stamped.
+//
+//	MessageImprint ::= SEQUENCE {
+//	 hashAlgorithm   AlgorithmIdentifier,
+//	 hashedMessage   OCTET STRING }
+type MessageImprint struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	HashedMessage []byte
+}
+
+// Request is a time-stamping request.
+//
+//	TimeStampReq ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 messageImprint  MessageImprint,
+//	 reqPolicy       TSAPolicyID              OPTIONAL,
+//	 nonce           INTEGER                  OPTIONAL,
+//	 certReq         BOOLEAN                  DEFAULT FALSE,
+//	 extensions      [0] IMPLICIT Extensions  OPTIONAL }
+type Request struct {
+	Version        int // fixed to 1 as defined in RFC 3161 2.4.1 Request Format
+	MessageImprint MessageImprint
+	ReqPolicy      asn1.ObjectIdentifier `asn1:"optional"`
+	Nonce          *big.Int              `asn1:"optional"`
+	CertReq        bool                  `asn1:"optional,default:false"`
+	Extensions     []pkix.Extension      `asn1:"optional,tag:0"`
+}
+
+// RequestOptions provides options for caller to create a new timestamp request
+type RequestOptions struct {
+	// Content is the datum to be time stamped. REQUIRED.
+	Content []byte
+
+	// HashAlgorithm is the hash algorithm to be used to hash the Content.
+	// REQUIRED and MUST be an available hash algorithm.
+	HashAlgorithm crypto.Hash
+
+	// HashAlgorithmParameters is the parameters for the HashAlgorithm.
+	// OPTIONAL.
+	HashAlgorithmParameters asn1.RawValue
+
+	// ReqPolicy specifies the TSA policy ID. OPTIONAL.
```

Comment:
Do you have any references to `ReqPolicy`?

Reply:
Added reference.